_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,231 @@
+for the build, and take the most recent one. Then it parses the file to
+see how many failed/completed tests there were.
+It does this for every build in the view, every time it is refreshed.
+This leads to unnecessary requests and computation. This, in turn, slows
+down the response times and increases the API load.
+
+Having the tests' error messages is also pretty much a necessity when trying
+to fix them. Currently, there is no way to ask the API for them.
+
+## Related GitHub issues
+
+- wharf-api [#11](https://github.com/iver-wharf/wharf-api/issues/11) [#17](https://github.com/iver-wharf/wharf-api/issues/17)
+- wharf-web [#13](https://github.com/iver-wharf/wharf-web/issues/13)
+
+## Explanation
+
+### wharf-api
+
+The `POST /build/{buildid}/artifact` endpoint handles inserting artifacts, same
+as before. However, if it's a test results file (eg: .trx), one should call the
+`POST /build/{buildid}/test-results` endpoint instead. This tells the server
+to parse them as well, creating several `TestResultDetail`s and one `TestResultSummary`
+per TRX file.
+
+The summaries get inserted into the database table `test_result_summary`.\
+The results get inserted into the database table `test_result_detail`.
+
+![Database structure](../_assets/wharf-db-graph.png)
+
+#### New or modified database models
+```diff
+// database_models.go
+type Build struct {
+  BuildID     uint         `gorm:"primaryKey" json:"buildId"`
+  StatusID    BuildStatus  `gorm:"not null" json:"statusId"`
+  ProjectID   uint         `gorm:"not null;index:build_idx_project_id" json:"projectId"`
+  Project     *Project     `gorm:"foreignKey:ProjectID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+  ScheduledOn *time.Time   `gorm:"nullable;default:NULL" json:"scheduledOn" format:"date-time"`
+  StartedOn   *time.Time   `gorm:"nullable;default:NULL" json:"startedOn" format:"date-time"`
+  CompletedOn *time.Time   `gorm:"nullable;default:NULL" json:"finishedOn" format:"date-time"`
+  GitBranch   string       `gorm:"size:300;default:'';not null" json:"gitBranch"`
+  Environment null.String  `gorm:"nullable;size:40" json:"environment" swaggertype:"string"`
+  Stage       string       `gorm:"size:40;default:'';not null" json:"stage"`
+  Params      []BuildParam `gorm:"foreignKey:BuildID" json:"params"`
+  IsInvalid   bool         `gorm:"not null;default:false" json:"isInvalid"` 
++ TestResultSummaries []TestResultSummary `gorm:"foreignKey:BuildID" json:"testResultSummaries"`
+}
+
++type TestResultSummary struct {
++  ArtifactID  uint      `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
++  Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
```

Comment:
Would personally prefer a CASCADE here instead of RESTRICT

```suggestion
+  Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"-"`
+  BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
+  Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"-"`
```

Docs: https://gorm.io/docs/constraints.html#Foreign-Key-Constraint & https://www.postgresql.org/docs/9.5/ddl-constraints.html#DDL-CONSTRAINTS-FK